divi/src/flat-database.h

```diff
@@ -170,80 +172,32 @@ class CFlatDB : private FlatDataFile
 
     ReadResult Read(T& objToLoad, const bool fDryRun = false) const
     {
-        //LOCK(objToLoad.cs);
-
         const int64_t nStart = GetTimeMillis();
-        // open input file, and associate with CAutoFile
-        FILE *file = fopen(Path().string().c_str(), "rb");
-        CAutoFile filein(file, SER_DISK, CLIENT_VERSION);
-        if (filein.IsNull())
-        {
-            error("%s: Failed to open file %s", __func__, Path().string());
+
+        auto reader = FlatDataFile::Read();
+        if (reader == nullptr)
             return FileError;
-        }
 
         // use file size to size memory buffer
         const int fileSize = boost::filesystem::file_size(Path());
         int dataSize = fileSize - sizeof(uint256);
         // Don't try to resize to a negative number if file is small
         if (dataSize < 0)
             dataSize = 0;
-        std::vector<unsigned char> vchData;
-        vchData.resize(dataSize);
-        uint256 hashIn;
 
-        // read data and checksum from file
-        try {
-            filein.read((char *)&vchData[0], dataSize);
-            filein >> hashIn;
-        } catch (const std::exception& e) {
-            error("%s: Deserialize or I/O error - %s", __func__, e.what());
-            return HashReadError;
-        }
-        filein.fclose();
```

Comment:
Closing the file was dropped from the new Reader class so the file remains open? Was this intentional seems like a potential source of error.

Reply:
`CAutoFile` auto-closes in the destructor, which gets called when the `Reader` instance is destructed.  If you think it helps with understandability of the code, I can of course do an explicit `file.close()` in a new destructor of `Reader`.

Reply:
No, that should be fine then. Thanks for pointing that out.